File: example/xx_three/lights_point.cpp
//#include <mud/frame.h>
#include <frame/Api.h>
#include <gfx-pbr/Api.h>

#include <xx_three/xx_three.h>

using namespace mud;

void xx_lights_point(Shell& app, Widget& parent, Dockbar& dockbar, bool init)
{
	UNUSED(dockbar);
	SceneViewer& viewer = ui::scene_viewer(parent);
	//ui::orbit_controls(viewer);
	TrackballController& controls = ui::trackball_controller(viewer);
	controls.m_dynamicDampingFactor = 0.15f;
	
	//controls.keys = [65, 83, 68];

	Scene& scene = viewer.m_scene;

	static vector<Node3*> lights;

	if(init)
	{
		Camera& camera = viewer.m_camera;
		camera.m_fov = 50.f; camera.m_near = 1.f; camera.m_far = 300.f;
		camera.m_eye = vec3(0.f, 15.f, 150.f);

		Texture& hdrenv = *app.m_gfx.textures().file("radiance/tiber_1_1k.hdr");
		scene.m_env.m_radiance.m_texture = &hdrenv;
		//scene.m_env.m_background.m_texture = &hdrenv;
		//scene.m_env.m_background.m_mode = BackgroundMode::Panorama;


		Material& groundmat = app.m_gfx.materials().create("ground", [&](Material& m) {
			m.m_program = &app.m_gfx.programs().fetch("pbr/pbr");
			m.m_base.m_uv0_scale = { 20.f, 10.f };
			m.m_pbr.m_albedo = rgb(0xffffff);
			m.m_pbr.m_albedo = rgb(0x444444);
			m.m_pbr.m_albedo = app.m_gfx.textures().file("disturb.jpg");
			// Phong
		});
		//new THREE.MeshPhongMaterial({ color: 0xffffff, map: texture });

		// GROUND

		Model& model = app.m_gfx.shape(Rect(vec2(0.f), vec2(800.f, 400.f)));
		Node3& node = gfx::nodes(scene).add(Node3(vec3(0.f, -5.f, 0.f), quat(vec3(0.f))));
		gfx::items(scene).add(Item(node, model, 0, &groundmat));

		// OBJECTS
		//Material& material = gfx::pbr_material(app.m_gfx, "object", Colour(1.f));
		Material& material = gfx::pbr_material(app.m_gfx, "object", rgb(0xffffff), 0.5f, 1.0f);
		//new THREE.MeshStandardMaterial({ color: 0xffffff, roughness: 0.5f, metalness: 1.0f });

		Model& torus = app.m_gfx.shape(Torus(1.5f, 0.4f));
		//Model& torus = app.m_gfx.shape(Torus(1.f, 0.1f));

		//for(int i = 0; i < 1; i++) {
		for(int i = 0; i < 5000; i++) {

			float x = 400.f * (0.5f - randf());
			float y = 50.f * (0.5f - randf()) + 25;
			float z = 200.f * (0.5f - randf());

			float a = 3.14f * (0.5f - randf());
			float b = 3.14f * (0.5f - randf());

			uint32_t flags = ItemFlag::Default;// | ItemFlag::NoCull;
			//Node3& n = gfx::nodes(scene).add(Node3(vec3(x, y, z), quat(a, b, 0, 1)));
			Node3& n = gfx::nodes(scene).add(Node3(vec3(x, y, z), quat(vec3(a, b, 0.f))));
			Item& it = gfx::items(scene).add(Item(n, torus, flags, &material));
			UNUSED(it);
		}

		// LIGHTS

		float intensity = 2.5f;
		float distance = 100.f;
		//float decay = 2.0f;

		uint32_t colours[] = { 0xff0040, 0x0040ff, 0x80ff80, 0xffaa00, 0x00ffaa, 0xff1100 };

		Sphere sphere = Sphere(0.25f); //THREE.SphereBufferGeometry(0.25, 16, 8);
		Model& sphere_model = app.m_gfx.shape(sphere);

		lights.clear();
		for(int i = 0; i < 6; ++i)
		{
			Colour c = rgb(colours[i]);
			Material& m = gfx::solid_material(app.m_gfx, ("light" + to_string(i)).c_str(), c); //Material({ color: colours[i] }));
			Node3& n = gfx::nodes(scene).add(Node3());
			Light& l = gfx::lights(scene).add(Light(n, LightType::Point, false, c, intensity, distance));
			l.m_attenuation = 2.f;
			//l.decay = decay;
			Item& it = gfx::items(scene).add(Item(n, sphere_model, ItemFlag::Default, &m));
			UNUSED(it);

			lights.push_back(&n);
		}

		//Node3& direct_node = gfx::nodes(scene) += Node3());
		//Light& direct_light = gfx::lights(scene) += Light(direct_node, LightType::Direct)); //THREE.DirectionalLight(0xffffff, 0.05);
		//dlight.position.set(0.5f, 1, 0).normalize();
	}

	float coef0[] = { 0.7f, 0.3f, 0.7f, 0.3f, 0.3f, 0.7f };
	float coef1[] = { 0.3f, 0.7f, 0.5f, 0.5f, 0.5f, 0.5f };

	float d = 150;

	const float time = app.m_gfx.m_time * 0.2f;

	for(int i = 0; i < 6; ++i)
	{
		vec2 pos = vec2(sin(time * coef0[i]) * d, cos(time * coef1[i]) * d);
		lights[i]->apply(vec3(pos.x, 0.f, pos.y));
	}
}

File: example/xx_three/two/lights_point.js

var viewer = two.ui.scene_viewer(panel);
var controls = two.ui.trackball_controller(viewer);
//controls.dynamicDampingFactor = 0.15;
//controls.keys = [ 65, 83, 68 ];

var scene = viewer.scene;

//scene.background = new THREE.Color( 0x040306 );
//scene.fog = new THREE.Fog( 0x040306, 10, 300 );

if (init) {
    //viewer.camera.eye = new two.vec3(0, 15, 150);
    
    var texture = app.gfx.textures.file('disturb.jpg');
    //texture.repeat.set(20, 10);

    var groundmat = app.gfx.materials.fetch('ground'); //new THREE.MeshPhongMaterial({ color: 0xffffff, map: texture });
    var material = two.gfx.pbr_material(app.gfx, 'object', two.rgb(0xffffff), 0.5, 1.0);

    // GROUND
    
    //var model = app.gfx.shape(new two.Rect(new two.vec2(800.0, 400.0));
    //var node = scene.nodes().add(new two.Node3(new two.vec3(0, -5, 0), new two.quat(-Math.PI / 2.0, 0, 0, 1)));
    //scene.items().add(new two.Item(node, model, 0, groundmat));

    // OBJECTS
    
    var torus = app.gfx.shape(new two.Torus(1.0, 0.1)); //new THREE.TorusBufferGeometry(1.5, 0.4, 8, 16);

    for (var i = 0; i < 5000; i ++) {

        var x = 400.0 * (0.5 - Math.random());
        var y = 50.0 * (0.5 - Math.random()) + 25;
        var z = 200.0 * (0.5 - Math.random());

        var a = 3.14 * (0.5 - Math.random());
        var b = 3.14 * (0.5 - Math.random());

        var n = scene.nodes().add(new two.Node3(new two.vec3(x, y, z), new two.quat(a, b, 0, 1)));
        var it = scene.items().add(new two.Item(n, torus, 0, material));
    }

    // LIGHTS

    var intensity = 2.5;
    var distance = 100;
    var decay = 2.0;

    var colours = [0xff0040, 0x0040ff, 0x80ff80, 0xffaa00, 0x00ffaa, 0xff1100];

    var sphere = app.gfx.shape(new two.Sphere(2.5));

    this.lights = [];

    for(var i = 0; i < 6; ++i)
    {
        var c = two.rgba(colours[i]);
        var m = two.gfx.solid_material(app.gfx, 'light' + i.toString(), c);
        var n = scene.nodes().add(new two.Node3());
        var l = scene.lights().add(new two.Light(n, two.LightType.Point, false, c, intensity, distance));
        l.attenuation = 2.0;
        //l.decay = decay;
        var it = scene.items().add(new two.Item(n, sphere, 0, m));
        
        this.lights.push(n);
    }

    //var direct_node = scene.nodes() += new two.Node3());
    //var direct_light = scene.lights() += Light(direct_node, two.LightType.Direct)); //THREE.DirectionalLight(0xffffff, 0.05);
    //dlight.position.set(0.5, 1, 0).normalize();
}

var root = viewer.scene.begin();
two.gfx.radiance(root, 'radiance/tiber_1_1k.hdr', two.BackgroundMode.Radiance);

var coef0 = [0.7, 0.3, 0.7, 0.3, 0.3, 0.7];
var coef1 = [0.3, 0.7, 0.5, 0.5, 0.5, 0.5];

//var time = Date.now() * 0.00025;
var d = 150;

for(var i = 0; i < 6; ++i)
{
    var pos = new two.vec3(Math.sin(time * coef0[i]) * d, 0.0, Math.cos(time * coef1[i]) * d);
    this.lights[i].apply(pos);
}
